### PR TITLE
Current NLP module v0.1.

### PR DESCRIPTION
**Pull request by mzhurovich**
Happy April Fools' Day! :stuck_out_tongue_winking_eye:

### CODE REVIEW

File: nlp/indent-all.sh [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+#mv grammar_begin.inl grammar_begin.h && make indent && mv grammar_begin.h grammar_begin.inl
```
Comment (dkorolev) [inline nlp/indent-all.sh]: Obsolete?

File: nlp/nlp.h [hunk 2]
```diff
@@ -0,0 +1,666 @@
...
+                size_t end,
+                std::function<void(const emitted_t&)> emit) const {
+    struct SignalUnitEmitted {};
+    try {
+      impl_.EvalImpl(query, begin, end, [&emit](const typename IMPL::emitted_t&) {
+        emit(Unit());
+        throw SignalUnitEmitted();
+      });
+    } catch (const SignalUnitEmitted&) {
+    }
```
Comment (dkorolev) [inline nlp/nlp.h]: A comment in this empty section would help.

File: nlp/nlp.h [hunk 3]
```diff
@@ -0,0 +1,666 @@
...
+  const LHS_IMPL& lhs_impl_;
+  const RHS_IMPL& rhs_impl_;
+  SeqImplUnitInRHS(const LHS_IMPL& impl, const RHS_IMPL& rhs_impl) : lhs_impl_(impl), rhs_impl_(rhs_impl) {}
+  void EvalImpl(const ::current::nlp::AnnotatedQuery<annotated_query_term_t>& query,
+                size_t begin,
+                size_t end,
+                std::function<void(const emitted_t&)> emit) const {
+    for (size_t i = begin; i <= end; ++i) {
+      // Rely on the fact that only a single `Unit` will be emitted.
+      // Start from testing the right hand side, as it's faster, and the order does not matter when a `Unit` present.
```
Comment (dkorolev) [inline nlp/nlp.h]: ... when a `Unit` *is* present.

File: nlp/nlp.h [hunk 4]
```diff
@@ -0,0 +1,666 @@
...
+};
+
+}  // namespace current::nlp::impl
+
+template <typename ANNOTATED_QUERY_TERM, class IMPL>
+struct SchemaBlock {
+  using annotated_query_term_t = ANNOTATED_QUERY_TERM;
+  using emitted_t = typename IMPL::emitted_t;
+
+  mutable std::string name_;  // `mutable` for `InjectName` to remain `const`.
```
Comment (dkorolev) [inline nlp/nlp.h]: :+1:

File: nlp/nlp.h [hunk 5]
```diff
@@ -0,0 +1,666 @@
...
+Optional<typename IMPL::emitted_t> JustMatchQuery(const SchemaBlock<ANNOTATED_QUERY_TERM, IMPL>& schema_block,
+                                                  S&& query_string) {
+  using emitted_t = typename IMPL::emitted_t;
+  const AnnotatedQuery<ANNOTATED_QUERY_TERM> query = AnnotateQuery<ANNOTATED_QUERY_TERM>(std::forward<S>(query_string));
+  Optional<emitted_t> return_value;
+  struct SignalEvalSucceeded {};
+  try {
+    schema_block.Eval(query, 0u, query.annotated_terms.size(), [&return_value](const emitted_t& result) {
+      return_value = result;
+      throw SignalEvalSucceeded();
```
Comment (dkorolev) [inline nlp/nlp.h]: A comment could help here as well.

File: nlp/nlp_schema_begin.inl [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+  static const name##_name_injector name##_injector_impl
+
+#define Void(term) ((term).UnitHelper())
+
+#define As(term, output_type)              \
+  ((term).template MapHelper<output_type>( \
+      [](const typename decltype(term)::emitted_t& input, output_type& output) { output = input; }))
+
+#define Map(term, output_type, ...)                                                                                  \
+  ((term).template MapHelper<output_type>([](const typename decltype(term)::emitted_t& input, output_type& output) { \
```
Comment (dkorolev) [inline nlp/nlp_schema_begin.inl]: Me wow.

File: nlp/test.cc [hunk 7]
```diff
@@ -0,0 +1,926 @@
...
+**********************************************************************************************************************/
+
+#include "nlp_schema_begin.inl"
+
+NLPSchema(DictionaryBasedAnnotation, C5TAnnotatedQueryTerm) {
+  CURRENT_STRUCT(CurrentMaintainer) {
+    CURRENT_FIELD(is_dima, bool);
+    CURRENT_FIELD(is_max, bool);
+    CURRENT_FIELD(is_grixa, bool);
+    // NOTE: Will need to implement default constructor generation as part of reflected schema output generation,
```
Comment (dkorolev) [inline nlp/test.cc]: Sigh. I could try to find a weekend for this.

**Comment by mzhurovich** [comment]
@dkorolev Thanks and PTAL!

**Pull request merged**